### ISSUE TITLE
Issue with DevExpress nuget feed

### ISSUE THREAD

**Issue by jcchalte** [fsprojects/Paket#2599]
Hi,

We have a project that use their products that we switched to paket three days ago, as a test before migrating our whole codebase. The migration was a blast, paket is VERY promising for our needs.

But we are encountering an issue with one of our supplier Nuget feed : DevExpress. DevExpress is a company that sells UI components for web & desktop applications. They usually package their components through an installer, but they started to distribute them through a private nuget server a few months ago. 

[https://www.devexpress.com/Support/Center/Question/Details/T466415/devexpress-nuget-packages](https://www.devexpress.com/Support/Center/Question/Details/T466415/devexpress-nuget-packages)

As we switched our package management system, we also tried to switch to the DevExpress nuget repository, but we cannot find a way to install the packages, as Paket seems to have issues with version numbers.

My paket.depencies file is as follows :

```
framework: auto-detect
redirects: on

source .\CustomNugetRepository
nuget EntityFramework = 6.2.0-alpha

source https://api.nuget.org/v3/index.json
nuget bootstrap 
...much more dependencies...
nuget WindowsAzure.Storage

source https://nuget.devexpress.com/MyPrivateKey/api
nuget DevExpress.Web 17.1.5.0
```

If I specify the DevExpress.Web latest version (17.1.5.0, or even `== 17.1.5.0`), `paket install` fails because it tries to resolve `DevExpress.Core 17.1.5` which does not exists on the server (the available version is '17.1.5.0'). If I specify an older version (`17.1.4.0`), I have another weird issue, paket does not seems to recognize one available version on the server

```
There was a version conflict during package resolution.
     Conflict detected:
      - Dependencies file requested package DevExpress.Web: == 17.1.4.0
      - Available versions:
        - (17.1.5.0, [https://nuget.devexpress.com/MyPrivateKey/api])
        - (17.1.4.0, [https://nuget.devexpress.com/MyPrivateKey/api])
        - (17.1.3.0, [https://nuget.devexpress.com/MyPrivateKey/api])
       ... And much more...
     Please try to relax some conditions or resolve the conflict manually (see http://fsprojects.github.io/Paket/nuget-dependencies.html#Use-exactly-this-version-constraint).
```

It might be an issue with the nuget server, but couldn't it be an issue with the way Paket manages version numbers with 4 parts ?

_I can obviously not expose publicly my private nuget key, but if this is required, I can send it privately._

**Comment by jcchalte** [comment]
As a side note, the Nuget client seems to work well. It seems to find the "wrong" version for all dependencies (`17.1.5` without the trailing `.0`), but it get the right version (`17.1.5.0`).

Here is the log (in French... sorry !)

```
Récupération du package 'DevExpress.Core 17.1.5' dans 'DevExpress'.
Récupération du package 'DevExpress.Data 17.1.5' dans 'DevExpress'.
Récupération du package 'DevExpress.Office.Core 17.1.5' dans 'DevExpress'.
Récupération du package 'DevExpress.PivotGrid.Core 17.1.5' dans 'DevExpress'.
Récupération du package 'DevExpress.RichEdit.Core 17.1.5' dans 'DevExpress'.
Récupération du package 'DevExpress.SpellChecker.Core 17.1.5' dans 'DevExpress'.
Récupération du package 'DevExpress.Web 17.1.5' dans 'DevExpress'.
  GET https://nuget.devexpress.com/MyKey/api/package/DevExpress.PivotGrid.Core/17.1.5.0
  GET https://nuget.devexpress.com/MyKey/api/package/DevExpress.Web/17.1.5.0
  GET https://nuget.devexpress.com/MyKey/api/package/DevExpress.Core/17.1.5.0
  GET https://nuget.devexpress.com/MyKey/api/package/DevExpress.Office.Core/17.1.5.0
  GET https://nuget.devexpress.com/MyKey/api/package/DevExpress.Data/17.1.5.0
  GET https://nuget.devexpress.com/MyKey/api/package/DevExpress.SpellChecker.Core/17.1.5.0
  GET https://nuget.devexpress.com/MyKey/api/package/DevExpress.RichEdit.Core/17.1.5.0
  OK https://nuget.devexpress.com/MyKey/api/package/DevExpress.PivotGrid.Core/17.1.5.0 203 ms
...
```

**Comment by forki** [comment]
can you please send the key to sforkmann at gmail?

Also why do you need to pin the package version in the deps? Couldn't you just use

    source https://nuget.devexpress.com/MyPrivateKey/api
    nuget DevExpress.Web

does that work?

**Comment by jcchalte** [comment]
If I don't pin it, it does not work either. Paket.exe outputs some exceptions as it cannot find the package or any dependency in our "CustomNugetRepository" (a simple folder), then tries every older version of the DevExpress package and finally enters in an infinite loop, output the same lines over and over again, without ever exiting the process :

```
DevExpress.Core 17.1.5
    Package not available.
      Message: Unable to retrieve package details for 'DevExpress.Core'-17.1.5
 - DevExpress.Core 17.1.5
    Package not available.
      Message: Unable to retrieve package details for 'DevExpress.Core'-17.1.5
 - DevExpress.Core 17.1.5
    Package not available.
      Message: Unable to retrieve package details for 'DevExpress.Core'-17.1.5
 - DevExpress.Core 17.1.5
    Package not available.
      Message: Unable to retrieve package details for 'DevExpress.Core'-17.1.5
...
```

**Comment by jcchalte** [comment]
Well, I stepped into the paket source code and found that there is something weird with the DevExpress server.

If I look at the OData feed returned by the DevExpress api (at https://nuget.devexpress.com/MyKey/api/Packages ), there is an entry for DevExpress.Web package that reference a DevExpress.Core dependency. But there is no DevExpress.Core entry in that document, thus preventing Paket from requesting this package details (in `NuGetV2.fs,`, in the `parseODataListDetails` function) as a filtered request to the OData feed with the "DevExpress.Core" ID returns nothing.


The `EmptyResult `is then carried back to `getPackageDetails `which return a null `maybePackageDetails`, which results in an exception.

But the package can be downloaded by following the same convention as all the others package, via one URL that is :

https://nuget.devexpress.com/MyKey/api/package/DevExpress.Core/17.1.5.0

So it seems that the exposed odata is not coherent. We can download a package, but we cannot ask details about it, including the download URL.

But as stated, Nuget seems to work fine, as it can download the DevExpress.Core package even though I cannot comprehend from where it retrieves the download URL... There is some sort of black magic at work here... Maybe the nuget source code can explain this mystery !

In conclusion, It seems that on some edge cases, Nuget is a little bit more loose than Paket on this topic. But it is an edge case that could be hard to correct on the Paket side. I will try to submit a ticket on the DevExpress bug system to see if they can correct it on their side.

**Comment by forki** [comment]
I'm sure there is a correct URL to get package details but paket doesn't
use it yet.
Maybe the devexpress people can answer that quickly

2017-08-10 17:38 GMT+02:00 Jean-Christophe Chalté <notifications@github.com>
:

**Comment by jcchalte** [comment]
I opened a public ticket [Here](https://www.devexpress.com/Support/Center/Question/Details/T544555/nuget-odata-feed-inconsistent-entries). We will see their answers.

**Comment by jcchalte** [comment]
In fact no, all the details are available by following the same convention. My bad, it was a mistake on my side.

Calling the URL (eg. https://nuget.devexpress.com/myKey/api/Packages(Id='DevExpress.Core',Version='17.1.5') ) retrieves all the expected details. So there must be an issue somewhere. Paket seems to struggle to find the right XML node in the GET result. I will try to find the issue.

**Comment by jcchalte** [comment]
Well, I seem to have found a way to make it work.

The culprit is that Paket first make an OData query to the API, trying to filter the packages by giving it a "where clause". But as the API does not expose publicly those packages in the root feed, it does not expose them either when this feed is filtered.

So, I added a fifth fallback method in the `NuGetV2.fs` `getDetailsFromNuGetViaODataFast` function that does not request the API through an Odata query but directly on the detail page, retrieving only one entry, and passing it to `parseODataEntryDetails`  method (instead of `parseODataListDetails`), and it seems to work quite well, on this case at least. I will try to clean my code as soon as possible, and send you a PR. I should also add a 6th fallback case to handle the "version vs normalizedVersion" if possible, if I understand it well.

**Comment by forki** [comment]
yes that's the correct way. This is what I would have done at the weekend.
If you send PR then this is less work for me - awesome!

2017-08-10 18:25 GMT+02:00 Jean-Christophe Chalté <notifications@github.com>
:

**Comment by matthid** [comment]
@jcchalte can you take a look at https://github.com/fsprojects/Paket/pull/2582
Ideally we would blacklist urls after we notice that they are not working, in a similar way as the pr suggests for getversion (but this time for getdetails). 

If it looks too complicated for you I'll take care of it eventually. Thanks for looking at this!

**Comment by matthid** [comment]
Just to be clear you pr is welcomed in any case!

**Comment by jcchalte** [comment]
Well I will look at #2582 if I can find the time :)

I think I can understand why DevExpress chose to not expose some packages "publicly". Those packages are present in order to share code between their products, but are not intended to be used by themselves. Exposing them as all the others would clutter the feed and the nuget user experience. Exposed packages are the only ones that most (if not all) customers would want to use.

I suspect that as Nuget handle it quite well, it could be a feature of Nuget that you can choose to expose only a fraction of all packages in the main feed.

**Comment by forki** [comment]
I think the answer is much simpler. They just implement a different OData
subset (like all other custom nuget servers do as well). Nuget just parses
them slightly different and we need to adapt like we always do in such
situations.

Am 10.08.2017 6:37 nachm. schrieb "Jean-Christophe Chalté" <
notifications@github.com>:

**Comment by forki** [comment]
fixed

**Comment by matthid** [comment]
Can you all please retry with  5.86.0-alpha002. Otherwise we might break you on next release.
